### PR TITLE
Reject native currency for non-native call

### PR DESCRIPTION
**Pull request by nishim3**
## Related Issue
Which issue does this pull request resolve?
#555 
## Description of changes
Added a require statement to make sure no ether is sent during a non-native currency type.

### CODE REVIEW

**Review by hensha256 — CHANGES_REQUESTED** [review]
So fast responding to that issue hahah. Just a couple of requests 🙏

File: src/PoolManager.sol [hunk 1]
```diff
@@ -267,6 +267,7 @@ contract PoolManager is IPoolManager, ProtocolFees, NoDelegateCall, ERC6909Claim
         if (currency.isNative()) {
             paid = msg.value;
         } else {
+            require(msg.value == 0, "msg.value should be 0 for non-native currency");
```
[the remark this answers is not in the document]
Comment (snreynolds) [reply src/PoolManager.sol:270]: Looks like you still have not defined a custom error. You'll need to declare it in `IPoolManager.sol` and then revert with it here. You can see examples of how we declare custom errors at the top of the `IPoolManager.sol` contract

**Review by snreynolds — CHANGES_REQUESTED** [review]
please add custom error thanks!

File: src/interfaces/IPoolManager.sol [hunk 2]
```diff
@@ -36,6 +36,9 @@ interface IPoolManager is IProtocolFees, IERC6909Claims {
     /// or on a pool that does not have a dynamic swap fee.
     error UnauthorizedDynamicSwapFeeUpdate();
 
+    ///@notice Thrown when native currency is passed to a non native settlement
+    error rejectNative();
```
Comment (hensha256) [inline src/interfaces/IPoolManager.sol:40]: Please can it start with a capital!
Maybe `NonZeroNativeValue`

**Review by hensha256 — CHANGES_REQUESTED** [review]
Requested a name change for the error.

Please can you add a test?
Maybe you can add a `PoolSettleTest.sol` that calls settle on the pool manager inside a lock. And then inside `PoolManager.t.sol` add `test_settle_revertsSendingNativeWithToken` and test that this new custom error is hit with `vm.expectRevert`

Its not easy, so if you would like me to do this testing let me know :)

**Comment by nishim3** [comment]
@hensha256 I have changed the name. If the issue is not at upmost priority, I would like to try to code the test. I will let you know if and when I face issues.

File: src/test/PoolSettleTest.sol [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+contract PoolSettleTest is PoolTestBase {
+    using CurrencyLibrary for Currency;
+    using SafeCast for uint256;
+
+    constructor(IPoolManager _manager) PoolTestBase(_manager) {}
+
+    struct CallbackData {
+        address sender;
+        PoolKey key;
+        uint256 amount0;
```
Comment (hensha256) [inline src/test/PoolSettleTest.sol:19]: what is the purpose of amount0 and amount1? i think you can remove them!

File: src/test/PoolSettleTest.sol [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+import {Currency, CurrencyLibrary} from "../types/Currency.sol";
+import {IPoolManager} from "../interfaces/IPoolManager.sol";
+import {PoolKey} from "../types/PoolKey.sol";
+import {PoolTestBase} from "./PoolTestBase.sol";
+import {SafeCast} from "../libraries/SafeCast.sol";
+
+contract PoolSettleTest is PoolTestBase {
+    using CurrencyLibrary for Currency;
+    using SafeCast for uint256;
```
Comment (hensha256) [inline src/test/PoolSettleTest.sol:12]: I dont think you need these 2 libraries?

**Review by hensha256 — COMMENTED** [review]
The test looks great!! Just a couple of comments

File: test/utils/Deployers.sol [hunk 5]
```diff
@@ -96,11 +100,12 @@ contract Deployers {
     function deployMintAndApprove2Currencies() internal returns (Currency, Currency) {
         MockERC20[] memory tokens = deployTokens(2, 2 ** 255);
 
-        address[6] memory toApprove = [
+        address[7] memory toApprove = [
             address(swapRouter),
             address(modifyLiquidityRouter),
             address(donateRouter),
             address(takeRouter),
+            address(settleRouter),
```
Comment (hensha256) [inline test/utils/Deployers.sol:108]: i dont think this is needed - youre not transferring tokens from the settleRouter

**Comment by nishim3** [comment]
Any idea why coverage is failing?

**Comment by hensha256** [comment]
> Any idea why coverage is failing?

Just because your PR is a fork that doesnt have access to Uniswap's github secrets :) its no problem

**Review by hensha256 — DISMISSED** [review]

**Comment by hensha256** [comment]
we just need @snreynolds to approve because she requested changes too
please can you merge in main? 🙏

**Review by snreynolds — APPROVED** [review]

**Pull request merged**